dbms/src/DataStreams/WindowBlockInputStream.cpp

```diff
@@ -272,12 +274,120 @@ Int64 WindowTransformAction::getPartitionEndRow(size_t block_rows)
     return left;
 }
 
+RowNumber WindowTransformAction::stepForward(const RowNumber & current_row, UInt64 n)
+{
+    auto dist = distance(current_row, partition_start);
+    assert(dist >= 0);
+    if (dist <= static_cast<Int64>(n))
+        return partition_start;
+
+    RowNumber result_row = current_row;
+
+    // The step happens only in a block
+    if (result_row.row >= n)
+    {
+        result_row.row -= n;
+        return result_row;
+    }
+
+    // The step happens between blocks
+    n -= (result_row.row + 1);
+    --result_row.block;
+    result_row.row = blockAt(result_row).rows - 1;
+    while (n > 0)
+    {
+        auto & block = blockAt(result_row);
+        if (block.rows > n)
+        {
+            result_row.row = block.rows - n - 1; // index, so we need to -1
+            break;
+        }
+        n -= block.rows;
+        --result_row.block;
+        result_row.row = blockAt(result_row).rows - 1;
+    }
+    return result_row;
+}
+
+std::tuple<RowNumber, bool> WindowTransformAction::stepBackward(const RowNumber & current_row, UInt64 n)
+{
+    // Range of rows is [frame_start, frame_end),
+    // and frame_end position is behind the position of the last frame row.
+    // So we need to ++n.
+    ++n;
+
+    auto dist = distance(partition_end, current_row);
+    assert(dist >= 1);
+
+    // Offset is too large and the partition_end is the longest position we can reach
+    if (dist <= static_cast<Int64>(n))
+        return std::make_tuple(partition_end, partition_ended);
+
+    // Now, frame_end is impossible to reach to partition_end.
+    RowNumber frame_end_row = current_row;
+    auto & block = blockAt(frame_end_row);
+
+    // The step happens only in a block
+    if ((block.rows - frame_end_row.row - 1) >= n)
+    {
+        frame_end_row.row += n;
+        return std::make_tuple(frame_end_row, partition_ended);
+    }
+
+    // The step happens between blocks
+    n -= block.rows - frame_end_row.row;
+    ++frame_end_row.block;
+    frame_end_row.row = 0;
+    while (n > 0)
+    {
+        auto block_rows = static_cast<Int64>(blockAt(frame_end_row).rows);
+        if (static_cast<Int64>(n) >= block_rows)
+        {
+            frame_end_row.row = 0;
+            ++frame_end_row.block;
+            n -= block_rows;
+            continue;
+        }
+
+        frame_end_row.row += n;
+        n = 0;
+    }
+
+    return std::make_tuple(frame_end_row, partition_ended);
+}
+
+Int64 WindowTransformAction::distance(RowNumber left, RowNumber right)
+{
+    if (left.block == right.block)
+        return left.row - right.row;
+
+    Int64 negative_sign = 1;
+
+    // Ensure that left is larger than right
+    if (left.block < right.block)
+    {
+        negative_sign = -1;
+        std::swap(left, right);
+    }
```

Comment:
It looks like we can just use `RUNTIME_CHECK(left > right)` and don't need to care negative case

Reply:
> It looks like we can just use `RUNTIME_CHECK(left > right)` and don't need to care negative case

I think sometimes the programme will pass parameters that left < right and we can not always guarantee that left is larger right. The swap executed here is just for easy implementation.

Reply:
ok